File: go/2024/01/solution.go
// Package solution contains solutions to part 1 and 2
package solution

import (
	"log"
	"sort"
	"strconv"
	"strings"
)

func parseInput(input string) (left []int, right []int) {
	// Remove whitespace
	lines := strings.Split(strings.Trim(string(input), "\n"), "\n")

	for _, line := range lines {
		stringSlice := strings.Fields(line)
		leftItem, err := strconv.Atoi(stringSlice[0])
		if err != nil {
			log.Fatalf("Couldn't convert left_item %v to int, %v", leftItem, err)
		}
		rightItem, err := strconv.Atoi(stringSlice[1])
		if err != nil {
			log.Fatalf("Couldn't convert right_item %v to int, %v", rightItem, err)
		}

		left = append(left, leftItem)
		right = append(right, rightItem)
	}

	return left, right
}

// Part1 returns the solution for day 1 part 1
func Part1(input string) int {
	answer := 0
	leftSlice, rightSlice := parseInput(input)

	sort.Slice(leftSlice, func(i, j int) bool {
		return leftSlice[i] < leftSlice[j]
	})
	sort.Slice(rightSlice, func(i, j int) bool {
		return rightSlice[i] < rightSlice[j]
	})

	for i := range leftSlice {
		left, right := leftSlice[i], rightSlice[i]
		if left < right {
			answer += right - left
		} else {
			answer += left - right
		}
	}

	return answer
}

// Part2 returns the solution for day 1 part 2
func Part2(input string) int {
	answer := 0
	leftSlice, rightSlice := parseInput(input)

	leftCount := map[int]int{}

	// Get unique values
	for _, v := range leftSlice {
		if _, ok := leftCount[v]; !ok {
			leftCount[v] = 0
		}
	}

	for key := range leftCount {
		for _, rightV := range rightSlice {
			if key == rightV {
				// Count occurrences
				leftCount[key]++
			}
		}
	}

	for _, v := range leftSlice {
		answer += v * leftCount[v]
	}

	return answer
}

File: python/2024/01/solution.py
from typing import Tuple


def parse_lists(input: str) -> Tuple[list, list]:
    lines = input.splitlines()
    a, b = [], []
    for line in lines:
        items = line.split()
        a.append(int(items[0]))
        b.append(int(items[1]))
    return a, b


def part_1(input: str) -> int:
    result = 0
    left, right = parse_lists(input)
    left.sort()
    right.sort()
    for index, _ in enumerate(left):
        result += abs(left[index] - right[index])
    return result


def get_item_count(item: int, item_count: dict) -> int:
    if item in item_count:
        return item_count[item]
    return 0


def get_occurrences(right: list) -> dict:
    result = {}
    for item in right:
        if item not in result:
            result[item] = 0
        result[item] += 1
    return result


def part_2(input: str) -> int:
    left_list, right_list = parse_lists(input)

    right_count = get_occurrences(right_list)

    result = 0
    for left_item in left_list:
        result += left_item * get_item_count(left_item, right_count)

    return result
